File: java/src/test/java/it/giannibombelli/wsc2026/e2e/support/E2EApiDriver.java
package it.giannibombelli.wsc2026.e2e.support;

import it.giannibombelli.wsc2026.booking.api.BookingResponse;
import it.giannibombelli.wsc2026.booking.api.PlaceBookingRequest;
import it.giannibombelli.wsc2026.giftcard.api.GiftCardResponse;
import it.giannibombelli.wsc2026.payment.api.PaymentDetailsResponse;
import it.giannibombelli.wsc2026.payment.api.TransactionResponse;
import it.giannibombelli.wsc2026.testsupport.JavalinTestHelper;
import tools.jackson.databind.ObjectMapper;

import java.math.BigDecimal;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;

public final class E2EApiDriver {

    private static final ObjectMapper JSON = new ObjectMapper();

    private final JavalinTestHelper javalin;

    public E2EApiDriver(JavalinTestHelper javalin) {
        this.javalin = javalin;
    }

    public GiftCardResponse issueGiftCard() {
        JavalinTestHelper.TestResponse response = javalin.post("/gift-cards");
        assertThat(response.status()).isEqualTo(201);
        GiftCardResponse giftCard = readValue(response, GiftCardResponse.class);
        assertThat(giftCard.id()).isNotNull();
        return giftCard;
    }

    public GiftCardResponse requestTopUp(UUID giftCardId, BigDecimal amount) {
        String body = String.format("{\"amount\": \"%s\"}", amount);
        JavalinTestHelper.TestResponse response = javalin.post("/gift-cards/" + giftCardId + "/top-up", body);
        assertThat(response.status()).isEqualTo(200);
        return readValue(response, GiftCardResponse.class);
    }

    public BookingResponse placeBooking(UUID giftCardId, BigDecimal amount) {
        PlaceBookingRequest request = new PlaceBookingRequest(amount, "E2E booking", giftCardId);
        String body = writeValue(request);
        JavalinTestHelper.TestResponse response = javalin.post("/bookings", body);
        assertThat(response.status()).isEqualTo(201);
        BookingResponse booking = readValue(response, BookingResponse.class);
        assertThat(booking.id()).isNotNull();
        assertThat(booking.giftCardId()).isEqualTo(giftCardId);
        return booking;
    }

    public TransactionResponse startFullPaymentTransaction(UUID paymentId, String provider, BigDecimal amount) {
        String body = String.format(
            "{\"provider\": \"%s\", \"providerReference\": \"%s\", \"amount\": \"%s\"}",
            provider,
            UUID.randomUUID(),
            amount
        );
        JavalinTestHelper.TestResponse response = javalin.post("/payments/" + paymentId + "/transactions", body);
        assertThat(response.status()).isEqualTo(202);
        return readValue(response, TransactionResponse.class);
    }

    public PaymentDetailsResponse findPaymentRequestedFor(String clientReference) {
        JavalinTestHelper.TestResponse response = javalin.get("/internals/payments?clientReference=" + clientReference);
        assertThat(response.status()).isEqualTo(200);
        PaymentDetailsResponse payment = readValue(response, PaymentDetailsResponse.class);
        assertThat(payment.status()).isEqualTo("REQUESTED");
        return payment;
    }

    public PaymentDetailsResponse getPaymentSummary(UUID paymentId) {
        JavalinTestHelper.TestResponse response = javalin.get("/payments/" + paymentId);
        assertThat(response.status()).isEqualTo(200);
        return readValue(response, PaymentDetailsResponse.class);
    }

    public GiftCardResponse getGiftCardSummary(UUID giftCardId) {
        JavalinTestHelper.TestResponse response = javalin.get("/gift-cards/" + giftCardId);
        assertThat(response.status()).isEqualTo(200);
        return readValue(response, GiftCardResponse.class);
    }

    public GiftCardResponse waitForGiftCardBalanceToEqual(UUID giftCardId, BigDecimal expectedBalance) {
        long deadline = System.currentTimeMillis() + 2000;
        while (System.currentTimeMillis() < deadline) {
            GiftCardResponse card = getGiftCardSummary(giftCardId);
            if (card.balance().compareTo(expectedBalance) == 0) {
                return card;
            }
            sleep(100);
        }
        throw new AssertionError("Gift card balance was not updated to " + expectedBalance + " within timeout");
    }

    public void assertPaymentMatches(PaymentDetailsResponse payment, UUID expectedId, String expectedClientReference, BigDecimal expectedAmount) {
        assertThat(payment.id()).isEqualTo(expectedId);
        assertThat(payment.clientReference()).isEqualTo(expectedClientReference);
        assertThat(payment.amount()).isEqualByComparingTo(expectedAmount);
        assertThat(payment.status()).isEqualTo("REQUESTED");
    }

    private <T> T readValue(JavalinTestHelper.TestResponse response, Class<T> clazz) {
        try {
            return JSON.readValue(response.body(), clazz);
        } catch (Exception e) {
            throw new RuntimeException("Failed to parse response body: " + response.body(), e);
        }
    }

    private String writeValue(Object value) {
        try {
            return JSON.writeValueAsString(value);
        } catch (Exception e) {
            throw new RuntimeException("Failed to serialize request body", e);
        }
    }

    private void sleep(long millis) {
        try {
            Thread.sleep(millis);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new RuntimeException(e);
        }
    }
}

File: typescript/tests/e2e/support/e2eApiDriver.ts
// Driver API per i test E2E: nasconde i dettagli HTTP e fornisce helper di dominio.

import type { GiftCardResponse } from '../../../src/giftcard/api/giftCardResponse.js';
import type { BookingResponse } from '../../../src/booking/api/bookingResponse.js';
import type { PaymentDetailsResponse } from '../../../src/payment/api/paymentDetailsResponse.js';
import type { ExpressTestHelper } from '../../testsupport/expressTestHelper.js';

function sleep(millis: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, millis));
}

export class E2EApiDriver {
  private readonly _helper: ExpressTestHelper;

  constructor(helper: ExpressTestHelper) {
    this._helper = helper;
  }

  async issueGiftCard(): Promise<GiftCardResponse> {
    const response = await this._helper.post('/gift-cards');
    expect(response.status).toBe(201);
    const giftCard = JSON.parse(response.body) as GiftCardResponse;
    expect(giftCard.id).not.toBeNull();
    return giftCard;
  }

  async requestTopUp(giftCardId: string, amount: number): Promise<GiftCardResponse> {
    const response = await this._helper.post(`/gift-cards/${giftCardId}/top-up`, { amount });
    expect(response.status).toBe(200);
    return JSON.parse(response.body) as GiftCardResponse;
  }

  async placeBooking(giftCardId: string, amount: number): Promise<BookingResponse> {
    const response = await this._helper.post('/bookings', {
      amount,
      description: 'E2E booking',
      giftCardId,
    });
    expect(response.status).toBe(201);
    const booking = JSON.parse(response.body) as BookingResponse;
    expect(booking.id).not.toBeNull();
    expect(booking.giftCardId).toBe(giftCardId);
    return booking;
  }

  async startFullPaymentTransaction(paymentId: string, provider: string, amount: number): Promise<PaymentDetailsResponse> {
    const response = await this._helper.post(`/payments/${paymentId}/transactions`, {
      provider,
      providerReference: crypto.randomUUID(),
      amount,
    });
    expect(response.status).toBe(202);
    return JSON.parse(response.body) as PaymentDetailsResponse;
  }

  async findPaymentRequestedFor(clientReference: string): Promise<PaymentDetailsResponse> {
    const response = await this._helper.get(`/internals/payments?clientReference=${encodeURIComponent(clientReference)}`);
    expect(response.status).toBe(200);
    const payment = JSON.parse(response.body) as PaymentDetailsResponse;
    expect(payment.status).toBe('REQUESTED');
    return payment;
  }

  async getPaymentSummary(paymentId: string): Promise<PaymentDetailsResponse> {
    const response = await this._helper.get(`/payments/${paymentId}`);
    expect(response.status).toBe(200);
    return JSON.parse(response.body) as PaymentDetailsResponse;
  }

  async getGiftCardSummary(giftCardId: string): Promise<GiftCardResponse> {
    const response = await this._helper.get(`/gift-cards/${giftCardId}`);
    expect(response.status).toBe(200);
    return JSON.parse(response.body) as GiftCardResponse;
  }

  async waitForGiftCardBalanceToEqual(giftCardId: string, expectedBalance: number): Promise<GiftCardResponse> {
    const deadline = Date.now() + 2000;
    while (Date.now() < deadline) {
      const card = await this.getGiftCardSummary(giftCardId);
      if (card.balance === expectedBalance) {
        return card;
      }
      await sleep(100);
    }
    throw new Error(`Gift card balance was not updated to ${expectedBalance} within timeout`);
  }

  assertPaymentMatches(
    payment: PaymentDetailsResponse,
    expectedId: string,
    expectedClientReference: string,
    expectedAmount: number
  ): void {
    expect(payment.id).toBe(expectedId);
    expect(payment.clientReference).toBe(expectedClientReference);
    expect(payment.amount).toBe(expectedAmount);
    expect(payment.status).toBe('REQUESTED');
  }
}
